Add 'usermap' option to authorisation module to avoid linked column in Users table

yahiakala opened the pull request:
As per #515 

Works like this:



    import anvil.server
    from anvil_extras.authorisation import authorisation_required
    from anvil_extras import authorisation

    # optional - default mode is "classic". Now supports "usermap"
    authorisation.set_mode("classic")

    @anvil.server.callable
    @authorisation_required("can_edit_sensitive_thing")
    def sensitive_server_function():
      do_stuff()

s-cork commented:
yes looks good.

I would consider a different api

the `has_permissions` function is only different in the first line:
how to get the row that defines roles
and defining how to get this row is all that is required.

If there was a way to set a config option like the `set_user_roles_getter` option, then the `has_permission` function would look more like

```python
def has_permission(user):
    if config.get_roles_row is None:
        user_roles_row = user
    else:
        user_roles_row = config.get_roles_row(user)
        
    ... # rest of function as before
```

the `set_user_roles_getter` could take various options and might look like...

```python

config = {
    "get_roles_row": None
}

def set_user_roles_getter(option):
    if option is None:
        config["get_roles_row"] = None
    elif callable(option):
        config["get_roles_row"] = None
    elif isinstance(option, str):
        config["get_roles_row"] = lambda user: getattr(app_tables, option).get(user=user)
    elif ...: # option is app_table
        config["get_roles_row"] = lambda user: option.get(user=user)
    else:
        raise TypeError(...)

```

and your use case would become

```python
authorisation.set_user_roles_getter(app_tables.user_map)
# or
authorisation.set_user_roles_getter("user_map")
```

above naming would need some thought... but i think it adds flexibility.

yahiakala commented:
I think I know what you're getting at. set_mode is probably too general with the potential for future, more granular customizations.

How about a set_config function that accepts a dict or keyword arguments? One of the keys could be 'get_roles_row'


    config = {
        "get_roles_row": None
    }

    def set_config(**kwargs):
        if "get_roles_row" in kwargs:
            set_user_roles_getter(kwargs["get_roles_row"])

    def set_user_roles_getter(option):
        if option is None:
            config["get_roles_row"] = None
        elif callable(option):
            config["get_roles_row"] = option
        elif isinstance(option, str):
            config["get_roles_row"] = lambda user: getattr(app_tables, option).get(user=user)
        elif ...: # option is app_table
            config["get_roles_row"] = lambda user: option.get(user=user)
        else:
            raise TypeError(...)

s-cork commented:
Yeah exactly.

yahiakala commented:
Made some changes for simplicity - currently only accepts a table name. More flexibility can be added later.

    def set_user_roles_getter(option):
        if option is None:
            config["get_roles_row"] = lambda user: app_tables.users.get(user=user)["roles"]
        elif isinstance(option, str):  # table name
            config["get_roles_row"] = lambda user: getattr(app_tables, option).get(
                user=user
            )["roles"]
        else:
            raise TypeError("set_user_roles_getter: option is not valid.")

In the has_permission:

    try:
        user_permissions = set(
            permission["name"]
            for role in config["get_roles_row"](user)
            for permission in role["permissions"]
        )
    except TypeError:
        return False

s-cork commented on server_code/authorisation.py:
```diff
@@ -38,7 +57,7 @@ def has_permission(permissions):
         user_permissions = set(
             permission["name"]
-            for role in user["roles"]
+            for role in config["get_roles_row"](user)
```
This will fail if the config option is never set

s-cork replied:
```suggestion
            for role in config["get_roles"](user)
```

s-cork reviewed (changes requested):
Unless I'm missing something.

s-cork commented on server_code/authorisation.py:
```diff
@@ -7,9 +7,28 @@
 
 __version__ = "2.6.0"
 
+config = {"get_roles_row": lambda user: app_tables.users.get(user=user)["roles"]}
```
This isn't right as the default option - the users table doesn't have a user column

s-cork commented on server_code/authorisation.py:
```diff
@@ -7,9 +7,28 @@
+
+def set_user_roles_getter(option):
+    if option is None:
+        config["get_roles_row"] = lambda user: app_tables.users.get(user=user)["roles"]
```
likewise

s-cork commented on docs/guides/modules/authorisation.rst:
```diff
@@ -22,7 +30,8 @@ Users and Permissions
 * Add entries to the permissions table. (e.g. 'can_view_stuff', 'can_edit_sensitive_thing')
 * Add entries to the roles table (e.g. 'admin') with links to the relevant permissions
-* In the Users table, link users to the relevant roles
+* Classic mode: In the Users table, link users to the relevant roles
+* Usermap mode: In the usermap table, link users to the relevant roles
```
docs need updating now

yahiakala commented:
Woops - fixed that. The default has been set to the user's table on line 14:

    config = {"get_roles_row": lambda user: app_tables.users.get(user=user)["roles"]}

s-cork reviewed (changes requested):
Can we also add something in the change log
also consider adding pre-commit to your work flow to avoid the pre-commit bot commits
see the contributing guide
might be worth cleaning up the git history if you're inclined

yahiakala commented:
Sorry about the mess :). I do my commits in the Anvil Editor but I'll tweak my process to make the relevant commits locally with pre-commit so my next PR will be much cleaner.

I fixed the code and made sure to test both default and custom mode. I also added an entry into the changelog and updated the docs.

s-cork commented on server_code/authorisation.py:
```diff
@@ -7,9 +7,28 @@
 
 __version__ = "2.6.0"
 
+config = {"get_roles_row": lambda user: anvil.users.get_user()["roles"]}
```
seems unnecessary to call `get_user()` since we already have it

```suggestion
config = {"get_roles_row": lambda user: user"roles"]}
```

s-cork commented on docs/guides/modules/authorisation.rst:
```diff
@@ -22,7 +30,8 @@ Users and Permissions
 * Add entries to the permissions table. (e.g. 'can_view_stuff', 'can_edit_sensitive_thing')
 * Add entries to the roles table (e.g. 'admin') with links to the relevant permissions
-* In the Users table, link users to the relevant roles
+* Classic mode: In the Users table, link users to the relevant roles
+* Custom mode: Set the table name where the user is linked to roles
```
remove this since it's no longer relevant
```suggestion
```

s-cork commented on docs/guides/modules/authorisation.rst:
```diff
@@ -52,6 +61,10 @@ function is called and raises an error if not:
 
     import anvil.server
     from anvil_extras.authorisation import authorisation_required
+    from anvil_extras import authorisation
+
+    # optional - set the table name with roles column (default: users table)
+    authorisation.set_config(get_roles_row="usermap")
```
```suggestion
```

move to configuration options example

s-cork commented on docs/guides/modules/authorisation.rst:
```diff
@@ -9,11 +9,19 @@ Installation
 You will need to setup the Users and Data Table services in your app:
 
   * Ensure that you have added the 'Users' service to your app
+
+Classic Mode (requires changes to Users table):
   * In the 'Data Tables' service, add:
   	* a table named 'permissions' with a text column named 'name'
 	* a table named 'roles' with a text column named 'name' and a 'link to table'column named 'permissions' that links to multiple rows of the permissions table
 	* a new 'link to table' column in the Users table named 'roles' that links to multiple rows of the 'roles' table
 
+Custom Mode (no changes to Users table):
+  * In the 'Data Tables' service, add:
+  	* a table named 'permissions' with a text column named 'name'
+	* a table named 'roles' with a text column named 'name' and a 'link to table'column named 'permissions' that links to multiple rows of the permissions table
+	* a table with a 'link to table' column named 'user' that links to a single row of the 'users' table and a 'link to table' column named 'roles' that links to multiple rows of the 'roles' table
+
```
```suggestion
  * In the 'Data Tables' service, add:
  	1. a table named 'permissions' with a text column named 'name'
	2. a table named 'roles' with a text column named 'name' and a 'link to table'column named 'permissions' that links to multiple rows of the permissions table
	3. a new 'link to table' column in the Users table named 'roles' that links to multiple rows of the 'roles' table (This behaviour is configurable - see below)

Configuration Options:
++++++++++++++++++++++

Certain default behaviour is configurable

By default the user's table expects a 'roles' column, that links to multiple rows of the 'roles' table.
Alternatively you can create a table that maps a user to linked rows of the role's table, avoiding adding extra columns to the user's table.

* Create a table called 'usermap' with a column 'user' and a column 'roles'.
* the 'user' column is a link to a row in the user's table. The 'roles' column is a link to multiple rows from the 'roles' table.

.. code-block:: python

    import anvil.server
    from anvil_extras.authorisation
    
    authorisation.set_config(get_roles="usermap")

Now the authorisation module will use the 'usermap' table to get roles for a user.


```

I'd maybe add the Configuration options just above API
And then add `set_config` to the API section

s-cork commented on server_code/authorisation.py:
```diff
@@ -7,9 +7,28 @@
+config = {"get_roles_row": lambda user: anvil.users.get_user()["roles"]}
+
+
+def set_config(**kwargs):
+    if "get_roles_row" in kwargs:
+        set_user_roles_getter(kwargs["get_roles_row"])
+
+
+def set_user_roles_getter(option):
```
let's make this private since it's not part of the public api

```suggestion
def set_config(**kwargs):
    if "get_roles" in kwargs:
        _set_user_roles_getter(kwargs["get_roles"])


def _set_user_roles_getter(option):
```

s-cork commented on server_code/authorisation.py:
```diff
@@ -7,9 +7,28 @@
+
+
+def set_user_roles_getter(option):
+    if option is None:
+        config["get_roles_row"] = lambda user: anvil.users.get_user()["roles"]
+    elif isinstance(option, str):  # table name
+        config["get_roles_row"] = lambda user: getattr(app_tables, option).get(
+            user=user
+        )["roles"]
```
```suggestion
    if option is None:
        config["get_roles"] = lambda user: user["roles"]
    elif isinstance(option, str):  # table name
        config["get_roles"] = lambda user: getattr(app_tables, option).get(
            user=user
        )["roles"]
```

s-cork reviewed (commented):
Some suggestions:
- I don't think classic/custom mode make sense in the docs anymore, I've made some suggestions there.
- let's add `set_config` to the API section of the docs.
- I think I prefer the config options `get_roles` over `get_roles_row`, which is cleaner and better describes what that function is doing. (It's actually getting a list or roles)
- amended default get_roles option

yahiakala commented:
Made all changes as requested

s-cork commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@
 ## Minor Changes
 * designer hints - add some designer hints to components
   https://github.com/anvilistas/anvil-extras/pull/512
+* authorisation - added config option to choose table name
```
```suggestion
* authorisation - added config option to avoid adding a roles column to the users table
```

s-cork commented on server_code/authorisation.py:
```diff
@@ -7,9 +7,28 @@
+            "roles"
+        ]
+    else:
+        raise TypeError("set_user_roles_getter: option is not valid.")
```
```suggestion
        raise TypeError("get_roles: option is not valid.")
```

s-cork reviewed (commented):
two more suggested tweaks then lgtm

yahiakala commented:
Done as well 👍

s-cork reviewed (approved):
LGTM - i'll leave @meatballs to give it a once over

Thanks @yahiakala for the contribution!

yahiakala commented:
@meatballs How's it look?

meatballs commented:
My previous review still stands. I don't like the use of lambdas.

s-cork commented:
@meatballs i don't think there was a previous review? I can't see one.

meatballs commented:
Oh, good grief. It's all been sitting there in 'pending' state and I hadn't noticed.

meatballs commented:
For a bit of context, Trey Hunner wrote a blog article a few years back that I wholeheartedly agree with:
https://treyhunner.com/2018/09/stop-writing-lambda-expressions/

s-cork commented:
merci @yahiakala are you happy to clean up the use of lambdas

s-cork commented:
Also you'll need to reset your branch, i've rebased and force pushed this branch after bumping anvil-extras to 2.6.1

s-cork commented:
ping @yahiakala

meatballs commented on server_code/authorisation.py:
```diff
@@ -7,9 +7,28 @@
 import functools
 
 import anvil.users
+from anvil.tables import app_tables
 
 __version__ = "2.6.1"
 
+config = {"get_roles": lambda user: user["roles"]}
+
+
+def set_config(**kwargs):
+    if "get_roles" in kwargs:
+        _set_user_roles_getter(kwargs["get_roles"])
+
+
+def _set_user_roles_getter(option):
+    if option is None:
+        config["get_roles"] = lambda user: user["roles"]
+    elif isinstance(option, str):  # table name
+        config["get_roles"] = lambda user: getattr(app_tables, option).get(user=user)[
+            "roles"
+        ]
```
```suggestion
import functools
from operator import itemgetter

import anvil.users
from anvil.tables import app_tables

__version__ = "2.6.1"

config = {"get_roles": itemgetter("roles")}


def set_config(**kwargs):
    if "get_roles" in kwargs:
        _set_user_roles_getter(kwargs["get_roles"])


def _get_roles_from_table(table_name, user):
    return getattr(app_tables, table_name).get(user=user)["roles"]


def _set_user_roles_getter(option):
    if option is None:
        config["get_roles"] = itemgetter("roles")
    elif isinstance(option, str):  # table name
        config["get_roles"] = functools.partial(_get_roles_from_table, option)
```
```

yahiakala commented:
Thank you kindly! I'll be turning on email notifications :)
